Added a Prefect example (#218)
DAGWorks-Inc/hamilton: examples/prefect/README.md

# Hamilton + Prefect

In this example, were going to show how to run Hamilton within a Prefect task.
- [**Prefect**](https://prefect.io) is an open source orchestrator written in Python. Its purpose is to launch and execute your workflows over your infrastructure. It has integrations with many popular tools (Snowflake, dbt, Ray, OpenAI, etc.)
- [**Hamilton**](https://github.com/dagworks-inc/hamilton) is a micro-framework to describe dataflows in Python. Its strength is expressing the flow of data & computation in a way that is straightforward to create, maintain, and resue (much like dbt for SQL)

## Why both?
Prefect includes useful features for production system such as job queues, caching, retries, logging, and more. However, each additional task you create comes with execution overhead and needs to be manually added to the DAG structure. Opting for less granular tasks ultimately reduces visibility and maintainability.

> In contrast, Hamilton automatically builds its DAG from the function definition, encouraging clean, small and single-purposed functions. By using Hamilton in a Prefect task, you gain this greater data lineage, reduce execution overhead, and get portable Python code you can run outside Prefect too.

Reviewer: ```suggestion
In contrast, Hamilton automatically builds its DAG from the function definition, encouraging clean, small and single-purposed functions. By using Hamilton in a Prefect task, you gain this greater insight into lineage, reduced prefect execution overhead, and you get portable Python code you can run outside Prefect too.
```

Author: Fixed in the latest commits

# Hamilton + Prefect

In this example, were going to show how to run a simple `data preprocessing -> model training -> model evaluation` workflow using Hamilton within Prefect tasks.

> - [**Prefect**](https://prefect.io) is an open source orchestrator written in Python. Its purpose is to launch and execute your workflows over your infrastructure. It has integrations with many popular tools (Snowflake, dbt, Ray, OpenAI, etc.)

Reviewer: ```suggestion
- [**Prefect**](https://prefect.io) is an open source orchestrator written in Python. Its purpose is to launch and execute your workflows over your infrastructure. 
```

> - [**Hamilton**](https://github.com/dagworks-inc/hamilton) is a micro-framework to describe dataflows in Python. Its strength is expressing the flow of data & computation in a way that is straightforward to create, maintain, and reuse (much like dbt for SQL)

Reviewer: ```suggestion
- [**Hamilton**](https://github.com/dagworks-inc/hamilton) is a micro-framework to describe dataflows in Python. Its strength is expressing the flow of data & computation in a way that is straightforward to create, maintain, and reuse (much like DBT has done for SQL). It is lightweight and can be run anywhere that python runs.
```

## Why both?

> Prefect includes useful features for production system such as job queues, caching, retries, logging, and more. However, each additional task you create comes with execution overhead and needs to be manually added to the DAG structure. Opting for less granular tasks ultimately reduces visibility and maintainability.

Reviewer: ```suggestion
Prefect is a heavy weight system that includes useful features for production system such as job queues, caching, retries, logging, and more. However, each additional "task" you create comes with execution overhead and needs to be manually added to the DAG structure. Opting for less granular tasks ultimately reduces visibility and maintainability, while opting for more creates more overhead.
```

> In contrast, Hamilton automatically builds its DAG from the function definition, encouraging clean, small and single-purposed functions. By using Hamilton in a Prefect task, you gain this greater insight into lineage, reduced Prefect execution overhead, and you get portable Python code you can run outside Prefect too.

Reviewer: ```suggestion
In contrast, Hamilton automatically builds its DAG from the function definition itself, encouraging clean, small and single-purposed functions. By using Hamilton in a Prefect task, you gain this greater insight into lineage, reduced Prefect execution overhead and a simpler DAG to manage, and you get portable Python code you can run outside Prefect too.
```

You'll notice that the Prefect workflow has 2 tasks: `prepare_data_task` and `train_and_evaluate_model_task` even though there are 3 function modules. While each function module best describes a logical grouping of functions, the Prefect tasks can help define how/where they should be executed. For example, the data processing task might require CPU resources while the model training and evaluation would benefit from having access to a GPU. This split allows us to benefit from Prefect's monitoring and tracking features, which make it easier to debug errors than having a single large Prefect task.

For another workflow, a large Hamilton DAG in a single Prefect task, or smaller Prefect tasks and Hamilton modules might have been preferred. This shows how the Hamilton `Driver` brings flexibility to where and how you run your code.

## Prefect setup

> The easiest way to get this example running is to sign up for Prefect's free tier and follow the [Prefect Cloud Quickstart](https://docs.prefect.io/latest/cloud/cloud-quickstart/) section. The steps are:

Reviewer: ```suggestion
The easiest way to get this example running is to sign up for Prefect's free tier and follow the [Prefect Cloud Quickstart](https://docs.prefect.io/latest/cloud/cloud-quickstart/) section. Don't want to sign up? Don't worry, as Prefect is open source, can instead opt to host and run everything yourself as described [here](https://docs.prefect.io/2.11.0/getting-started/installation/). The steps are:
```
